File: app/schemas/parent.py
from pydantic import BaseModel, EmailStr, Field, field_validator
from datetime import date, datetime
from typing import Optional

from app.schemas.course import CourseResponse
from app.schemas.assignment import AssignmentResponse
from app.schemas.user import strip_whitespace


class CreateChildRequest(BaseModel):
    full_name: str = Field(min_length=1, max_length=255)
    email: EmailStr | None = None
    relationship_type: str = Field(default="guardian", max_length=50)

    @field_validator('full_name', mode='before')
    @classmethod
    def _strip_whitespace(cls, v: object) -> object:
        return strip_whitespace(v)


class LinkChildRequest(BaseModel):
    student_email: EmailStr
    full_name: str | None = Field(default=None, max_length=255)
    relationship_type: str = Field(default="guardian", max_length=50)

    @field_validator('full_name', mode='before')
    @classmethod
    def _strip_whitespace(cls, v: object) -> object:
        return strip_whitespace(v)


class ChildSummary(BaseModel):
    student_id: int
    user_id: int
    full_name: str
    email: Optional[str] = None
    grade_level: Optional[int]
    school_name: Optional[str]
    date_of_birth: Optional[date] = None
    phone: Optional[str] = None
    address: Optional[str] = None
    city: Optional[str] = None
    province: Optional[str] = None
    postal_code: Optional[str] = None
    notes: Optional[str] = None
    interests: list[str] = []
    relationship_type: str | None = None
    invite_link: str | None = None
    link_request_pending: bool = False
    course_count: int = 0
    active_task_count: int = 0
    invite_status: str | None = None  # "active", "pending", or "email_unverified"
    invite_id: int | None = None  # ID of pending invite (for resend)

    class Config:
        from_attributes = True


class DiscoveredChild(BaseModel):
    user_id: int
    email: str
    full_name: str
    google_courses: list[str]
    already_linked: bool


class DiscoverChildrenResponse(BaseModel):
    discovered: list[DiscoveredChild]
    google_connected: bool
    courses_searched: int


class LinkChildrenBulkRequest(BaseModel):
    user_ids: list[int]
    relationship_type: str = Field(default="guardian", max_length=50)


class CourseWithTeacher(CourseResponse):
    teacher_name: str | None = None
    teacher_email: str | None = None


class ChildUpdateRequest(BaseModel):
    full_name: Optional[str] = Field(default=None, max_length=255)
    email: Optional[EmailStr] = None
    grade_level: Optional[int] = None
    school_name: Optional[str] = Field(default=None, max_length=200)
    date_of_birth: Optional[date] = None
    phone: Optional[str] = Field(default=None, max_length=20)
    address: Optional[str] = Field(default=None, max_length=500)
    city: Optional[str] = Field(default=None, max_length=100)
    province: Optional[str] = Field(default=None, max_length=100)
    postal_code: Optional[str] = Field(default=None, max_length=20)
    notes: Optional[str] = Field(default=None, max_length=2000)
    interests: Optional[list[str]] = None

    @field_validator('interests', mode='before')
    @classmethod
    def validate_interests(cls, v: list[str] | None) -> list[str] | None:
        if v is None:
            return None
        if len(v) > 10:
            raise ValueError("Maximum 10 interests allowed")
        cleaned = []
        for item in v:
            if not isinstance(item, str):
                raise ValueError("Each interest must be a string")
            stripped = item.strip().lower()
            if not stripped:
                continue
            if len(stripped) > 50:
                raise ValueError("Each interest must be 50 characters or less")
            cleaned.append(stripped)
        return cleaned

    @field_validator('full_name', 'school_name', 'address', 'city', 'province', 'notes', mode='before')
    @classmethod
    def _strip_whitespace(cls, v: object) -> object:
        return strip_whitespace(v)


class ChildOverview(BaseModel):
    student_id: int
    user_id: int
    full_name: str
    grade_level: Optional[int]
    google_connected: bool = False
    courses: list[CourseWithTeacher]
    assignments: list[AssignmentResponse]
    study_guides_count: int


class ChildHighlight(BaseModel):
    """Per-child summary for the parent dashboard status view."""
    student_id: int
    user_id: int
    full_name: str
    grade_level: Optional[int] = None
    overdue_count: int = 0
    due_today_count: int = 0
    upcoming_count: int = 0
    completed_today_count: int = 0
    courses: list[CourseWithTeacher] = []
    overdue_items: list[dict] = []
    due_today_items: list[dict] = []


class ChildResetPasswordRequest(BaseModel):
    new_password: str | None = Field(default=None, min_length=8, max_length=128)


class LinkTeacherRequest(BaseModel):
    teacher_email: EmailStr
    teacher_name: str | None = Field(default=None, max_length=255)

    @field_validator('teacher_name', mode='before')
    @classmethod
    def _strip_whitespace(cls, v: object) -> object:
        return strip_whitespace(v)


class LinkedTeacher(BaseModel):
    id: int
    student_id: int
    teacher_user_id: int | None = None
    teacher_name: str | None = None
    teacher_email: str | None = None
    added_by_user_id: int
    created_at: datetime | None = None

    class Config:
        from_attributes = True


class ParentDashboardResponse(BaseModel):
    """Aggregated dashboard data returned in a single API call."""
    children: list[ChildSummary]
    google_connected: bool = False
    unread_messages: int = 0
    total_overdue: int = 0
    total_due_today: int = 0
    total_tasks: int = 0
    child_highlights: list[ChildHighlight] = []
    all_assignments: list[AssignmentResponse] = []
    all_tasks: list[dict] = []

File: ClassBridgeMobile/src/api/parent.ts
import { api } from './client';

// Parent Types (ported from web frontend/src/api/parent.ts)
export interface ChildHighlight {
  student_id: number;
  user_id: number;
  full_name: string;
  grade_level: number | null;
  overdue_count: number;
  due_today_count: number;
  upcoming_count: number;
  completed_today_count: number;
  courses: Array<{
    id: number;
    name: string;
    description: string | null;
    subject: string | null;
    google_classroom_id: string | null;
    teacher_id: number | null;
    created_at: string;
    teacher_name: string | null;
    teacher_email: string | null;
  }>;
  overdue_items: Array<{
    title: string;
    type: string;
    course_name: string;
    due_date: string;
  }>;
  due_today_items: Array<{
    title: string;
    type: string;
    course_name: string;
    due_date: string;
  }>;
}

export interface ParentDashboardData {
  children: ChildSummary[];
  google_connected: boolean;
  unread_messages: number;
  total_overdue: number;
  total_due_today: number;
  total_tasks: number;
  child_highlights: ChildHighlight[];
  all_assignments: Array<{
    id: number;
    title: string;
    description: string | null;
    course_id: number;
    google_classroom_id: string | null;
    due_date: string | null;
    max_points: number | null;
    created_at: string;
  }>;
  all_tasks: Array<Record<string, unknown>>;
}

export interface ChildSummary {
  student_id: number;
  user_id: number;
  full_name: string;
  email: string | null;
  grade_level: number | null;
  school_name: string | null;
  date_of_birth: string | null;
  phone: string | null;
  address: string | null;
  city: string | null;
  province: string | null;
  postal_code: string | null;
  notes: string | null;
  relationship_type: string | null;
  invite_link: string | null;
  course_count: number;
  active_task_count: number;
}

export interface ChildOverview {
  student_id: number;
  user_id: number;
  full_name: string;
  grade_level: number | null;
  google_connected: boolean;
  courses: Array<{
    id: number;
    name: string;
    description: string | null;
    subject: string | null;
    google_classroom_id: string | null;
    teacher_id: number | null;
    created_at: string;
    teacher_name: string | null;
    teacher_email: string | null;
  }>;
  assignments: Array<{
    id: number;
    title: string;
    description: string | null;
    course_id: number;
    google_classroom_id: string | null;
    due_date: string | null;
    max_points: number | null;
    created_at: string;
  }>;
  study_guides_count: number;
}

// Parent API — mobile uses read-only subset
export const parentApi = {
  getDashboard: async () => {
    const response = await api.get('/api/parent/dashboard');
    return response.data as ParentDashboardData;
  },

  getChildren: async () => {
    const response = await api.get('/api/parent/children');
    return response.data as ChildSummary[];
  },

  getChildOverview: async (studentId: number) => {
    const response = await api.get(
      `/api/parent/children/${studentId}/overview`
    );
    return response.data as ChildOverview;
  },
};
